x/avs/keeper/keeper.go

```diff
@@ -44,47 +46,83 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 	operatorAddress := params.OperatorAddress
 	opAccAddr, err := sdk.AccAddressFromBech32(operatorAddress)
 	if err != nil {
 		return errorsmod.Wrap(err, fmt.Sprintf("error occurred when parse acc address from Bech32,the addr is:%s", operatorAddress))
 	}
 	if !k.operatorKeeper.IsOperator(ctx, opAccAddr) {
-		return errorsmod.Wrap(delegationtypes.ErrOperatorNotExist, "AVSInfoUpdate: invalid operator address")
-	}
-
-	action := params.Action
-
-	if action == RegisterAction {
-		return k.SetAVSInfo(ctx, params.AvsName, params.AvsAddress, operatorAddress, params.AssetID)
+		return errorsmod.Wrap(delegationtypes.ErrOperatorNotExist, fmt.Sprintf("AVSInfoUpdate: invalid operator address:%s", operatorAddress))
 	}
 	avsInfo, err := k.GetAVSInfo(ctx, params.AvsAddress)
-	if err != nil {
+	if err != nil || avsInfo.GetInfo() == nil {
 		return errorsmod.Wrap(err, fmt.Sprintf("error occurred when get avs info %s", avsInfo))
 	}
-	if avsInfo.Info.AvsOwnerAddress != params.AvsOwnerAddress {
-		return errorsmod.Wrap(err, fmt.Sprintf("not qualified to deregister %s", avsInfo))
+	avs := avsInfo.GetInfo()
+	addresses := avs.OperatorAddress
+
+	if params.Action == RegisterAction && slices.Contains(avs.OperatorAddress, operatorAddress) {
+		return errorsmod.Wrap(types.ErrAlreadyRegistered, fmt.Sprintf("Error: Already registered，operatorAddress %s", operatorAddress))
 	}
-	return k.DeleteAVSInfo(ctx, params.AvsAddress)
+
+	if params.Action == RegisterAction && !slices.Contains(avs.OperatorAddress, operatorAddress) {
+		addresses = append(addresses, operatorAddress)
+		avs.OperatorAddress = addresses
+		return k.SetAVSInfo(ctx, avs)
+	}
+
+	if params.Action == DeRegisterAction && slices.Contains(avs.OperatorAddress, operatorAddress) {
+		avs.OperatorAddress = types.RemoveOperatorAddress(addresses, operatorAddress)
+		return k.SetAVSInfo(ctx, avs)
+	}
+	return errorsmod.Wrap(types.ErrUnregisterNonExistent, fmt.Sprintf("No available operatorAddress to DeRegisterAction ,operatorAddress: %s", operatorAddress))
 }
 
-func (k Keeper) SetAVSInfo(ctx sdk.Context, avsName, avsAddress, operatorAddress, assetID string) (err error) {
-	avsAddr, err := sdk.AccAddressFromBech32(avsAddress)
+func (k Keeper) SetAVSInfo(ctx sdk.Context, avs *types.AVSInfo) (err error) {
+	avsAddr, err := sdk.AccAddressFromBech32(avs.AvsAddress)
 	if err != nil {
 		return errorsmod.Wrap(err, "SetAVSInfo: error occurred when parse acc address from Bech32")
 	}
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixAVSInfo)
-	if !store.Has(avsAddr) {
-		AVSInfo := &types.AVSInfo{Name: avsName, AvsAddress: avsAddress, OperatorAddress: []string{operatorAddress}, AssetId: []string{assetID}}
-		bz := k.cdc.MustMarshal(AVSInfo)
-		store.Set(avsAddr, bz)
-		return nil
-	}
-	value := store.Get(avsAddr)
-	ret := &types.AVSInfo{}
-	k.cdc.MustUnmarshal(value, ret)
-	ret.OperatorAddress = append(ret.OperatorAddress, operatorAddress)
-	ret.AssetId = append(ret.AssetId, assetID)
+
+	AVSInfo := avs
```

Comment:
this AVSInfo  variable is the same as types.AVSInfo

Reply:
fixed